### PR TITLE
Override Foldable methods

### PR DESCRIPTION
**Pull request by peterneyens**
- Override `Foldable` methods in a lot of instances (including `Reducible` and `NonEmptyReducible`)
- Override `MonoidK#algebra` for `List`, `Vector`, etc to get kernel `Monoid` instance
- Add `Reducible` for `Tuple2`

I might have gone a bit overboard with overriding in the "one element `Foldable`" instances (`Option`, `Either`, ...).

### CODE REVIEW

File: core/src/main/scala/cats/Reducible.scala [hunk 1]
```diff
@@ -210,4 +220,55 @@ abstract class NonEmptyReducible[F[_], G[_]](implicit G: Foldable[G]) extends Re
...
+    val (a, ga) = split(fa)
+    p(a) && G.forall(ga)(p)
+  }
+
+  override def toList[A](fa: F[A]): List[A] = {
+    val (a, ga) = split(fa)
+    a :: G.toList(ga)
+  }
+
+  override def toNonEmptyList[A](fa: F[A]): NonEmptyList[A] = {
```
Comment (johnynek) [inline core/src/main/scala/cats/Reducible.scala]: this method code path is untested

File: core/src/main/scala/cats/data/NonEmptyVector.scala [hunk 2]
```diff
@@ -251,6 +250,21 @@ private[data] sealed trait NonEmptyVectorInstances {
         NonEmptyVector.fromVectorUnsafe(buf.result())
       }
 
+      override def fold[A](fa: NonEmptyVector[A])(implicit A: Monoid[A]): A =
+        fa.reduce
+
+      override def foldM[G[_], A, B](fa: NonEmptyVector[A], z: B)(f: (B, A) => G[B])(implicit G: Monad[G]): G[B] =
+        Foldable.iteratorFoldM(fa.toVector.toIterator, z)(f)
```
Comment (johnynek) [inline core/src/main/scala/cats/data/NonEmptyVector.scala]: code path is untested.

File: core/src/main/scala/cats/data/Validated.scala [hunk 3]
```diff
@@ -323,7 +324,39 @@ private[data] sealed abstract class ValidatedInstances extends ValidatedInstance
           case v @ Validated.Valid(_) => v
         }
       def raiseError[A](e: E): Validated[E, A] = Validated.Invalid(e)
+
+      override def reduceLeftToOption[A, B](fa: Validated[E, A])(f: A => B)(g: (B, A) => B): Option[B] =
```
Comment (johnynek) [inline core/src/main/scala/cats/data/Validated.scala]: next 5 methods untested.

File: core/src/main/scala/cats/data/Validated.scala [hunk 4]
```diff
@@ -323,7 +324,39 @@ private[data] sealed abstract class ValidatedInstances extends ValidatedInstance
...
+        fa.toOption.filter(f)
+
+      override def exists[A](fa: Validated[E, A])(p: A => Boolean): Boolean =
+        fa.exists(p)
+
+      override def forall[A](fa: Validated[E, A])(p: A => Boolean): Boolean =
+        fa.forall(p)
+
+      override def toList[A](fa: Validated[E, A]): List[A] =
+        fa.fold(_ => Nil, _ :: Nil)
```
Comment (johnynek) [inline core/src/main/scala/cats/data/Validated.scala]: untested

File: core/src/main/scala/cats/instances/either.scala [hunk 5]
```diff
@@ -83,12 +83,47 @@ trait EitherInstances extends cats.kernel.instances.EitherInstances {
...
       override def recover[B](fab: Either[A, B])(pf: PartialFunction[A, B]): Either[A, B] =
         fab recover pf
+
       override def recoverWith[B](fab: Either[A, B])(pf: PartialFunction[A, Either[A, B]]): Either[A, B] =
         fab recoverWith pf
+
       override def ensure[B](fab: Either[A, B])(error: => A)(predicate: B => Boolean): Either[A, B] =
         fab.ensure(error)(predicate)
+
+      override def reduceLeftToOption[B, C](fab: Either[A, B])(f: B => C)(g: (C, B) => C): Option[C] =
```
Comment (johnynek) [inline core/src/main/scala/cats/instances/either.scala]: next 5 method untested.

File: core/src/main/scala/cats/instances/either.scala [hunk 6]
```diff
@@ -83,12 +83,47 @@ trait EitherInstances extends cats.kernel.instances.EitherInstances {
...
+        fab.fold(_ => None, r => if (f(r)) Some(r) else None)
+
+      override def exists[B](fab: Either[A, B])(p: B => Boolean): Boolean =
+        fab.right.exists(p)
+
+      override def forall[B](fab: Either[A, B])(p: B => Boolean): Boolean =
+        fab.right.forall(p)
+
+      override def toList[B](fab: Either[A, B]): List[B] =
+        fab.fold(_ => Nil, _ :: Nil)
```
Comment (johnynek) [inline core/src/main/scala/cats/instances/either.scala]: untested.

File: core/src/main/scala/cats/instances/option.scala [hunk 7]
```diff
@@ -71,12 +71,46 @@ trait OptionInstances extends cats.kernel.instances.OptionInstances {
       override def filter[A](fa: Option[A])(p: A => Boolean): Option[A] =
         fa.filter(p)
 
+      override def reduceLeftToOption[A, B](fa: Option[A])(f: A => B)(g: (B, A) => B): Option[B] =
```
Comment (johnynek) [inline core/src/main/scala/cats/instances/option.scala]: next 5 methods untested.

File: core/src/main/scala/cats/instances/option.scala [hunk 8]
```diff
@@ -71,12 +71,46 @@ trait OptionInstances extends cats.kernel.instances.OptionInstances {
...
+
       override def exists[A](fa: Option[A])(p: A => Boolean): Boolean =
         fa.exists(p)
 
       override def forall[A](fa: Option[A])(p: A => Boolean): Boolean =
         fa.forall(p)
 
+      override def toList[A](fa: Option[A]): List[A] = fa.toList
+
+      override def filter_[A](fa: Option[A])(p: A => Boolean): List[A] =
```
Comment (johnynek) [inline core/src/main/scala/cats/instances/option.scala]: next 5 methods untested.

**Comment by peterneyens** [comment]
Improved the test coverage, only the methods `reduceLeftToOption` and `reduceRightToOption` are not yet tested.

File: core/src/main/scala/cats/instances/list.scala [hunk 9]
```diff
@@ -86,6 +86,19 @@ trait ListInstances extends cats.kernel.instances.ListInstances {
...
+
+      override def find[A](fa: List[A])(f: A => Boolean): Option[A] = fa.find(f)
+
+      override def filter_[A](fa: List[A])(p: A => Boolean): List[A] = fa.filter(p)
+
+      override def takeWhile_[A](fa: List[A])(p: A => Boolean): List[A] = fa.takeWhile(p)
+
+      override def dropWhile_[A](fa: List[A])(p: A => Boolean): List[A] = fa.dropWhile(p)
+
+      override def algebra[A]: Monoid[List[A]] = new kernel.instances.ListMonoid[A]
```
Comment (kailuowang) [inline core/src/main/scala/cats/instances/list.scala]: should it be `val` or `lazy val`?

**Comment by kailuowang** [comment]
I think the code is reasonable, especially now that we have full test coverage. 
I just want to share my thoughts. To me, whether to override a default implementation provided by the type class is a trade-off between performance and complexity. The complexity cost comes in in the sense of extra code to maintain and test against, and in some cases, extra laws.  Very often the performance gain outweighs the complexity cost. I feel that it's a pity we don't have a really good gauge on some of the performance gain here, e.g. the one element `Foldable`s. On the other hand, the complexity cost isn't very substantial either. 
For now, I think this PR is good for my 👍. 
Hopefully, in the future, we'll have some benchmarking to give us more data when making such performance/complexity trade-off decisions.

**Comment by peterneyens** [comment]
I concur that some benchmarking would be nice to make these performance/complexity decisions.

**Comment by kailuowang** [comment]
merging with two thumb ups.

**Pull request merged**